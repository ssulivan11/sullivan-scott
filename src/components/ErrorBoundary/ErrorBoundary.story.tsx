import React, { Fragment, useState } from 'react'
import { storiesOf } from '@storybook/react'

import ErrorBoundary from './ErrorBoundary'

const stories = storiesOf('Components | Error Boundary', module)

function getItem(type = 'default') {
  const demoStyle = { padding: 15 }
  const containerStyle = { marginTop: 25, maxWidth: 600, padding: 15, borderRadius: 5, backgroundColor: '#f9f9f9' }
  return (
    <div>
      <h5 style={demoStyle}>
        Increasing any counter to the magic number of <strong>3</strong> will simulate a JS error in the component.
      </h5>
      <div>
        {type === 'default' && (
          <div data-test='error-boundary-top-level' style={demoStyle}>
            <h3>Top Level</h3>
            <p className='no-line-height'>These three components are inside the same error top level boundary.</p>
            <p className='no-line-height'>If one crashes, the error boundary will replace both of them.</p>
            <div className='container' style={containerStyle}>
              <strong>Page Demo Example:</strong>
              <ErrorBoundary showErrorMsg>
                <BuggyCounter component='header' />
                <BuggyCounter component='body' />
                <BuggyCounter component='footer' />
              </ErrorBoundary>
            </div>
          </div>
        )}

        {type === 'retries' && (
          <div data-test='error-boundary-top-level-retries' style={demoStyle}>
            <h3>Top Level w/ 2 Retries</h3>
            <p className='no-line-height'>These three components are inside the same error top level boundary.</p>
            <p className='no-line-height'>
              There is a retry limit set to 2 in this case. So the boundary can fail only a max of 2 times.
            </p>
            <div className='container' style={containerStyle}>
              <ErrorBoundary showErrorMsg retries={2}>
                <BuggyCounter component='header' />
                <BuggyCounter component='body' />
                <BuggyCounter component='footer' />
              </ErrorBoundary>
            </div>
          </div>
        )}

        {type === 'component-level' && (
          <div data-test='error-boundary-component-level' style={demoStyle}>
            <h3>HOC Level</h3>
            <p className='no-line-height'>These three components are each inside of their own error boundary.</p>
            <p className='no-line-height'>So if one crashes, the other is not affected.</p>
            <div className='container' style={containerStyle}>
              <ErrorBoundary>
                <BuggyCounter component='header - no error msg' />
              </ErrorBoundary>
              <ErrorBoundary showErrorMsg>
                <BuggyCounter component='body' />
              </ErrorBoundary>
              <ErrorBoundary showErrorMsg errorMsg='Looks like the footer is having an issue.'>
                <BuggyCounter component='footer - unique error' />
              </ErrorBoundary>
            </div>
          </div>
        )}

        {type === 'entry-error' && (
          <div style={demoStyle}>
            <h3>JS Entry Error</h3>
            <p className='no-line-height'>
              As component renders, there is a JS error. The prop showErrorMsg is set to true.
            </p>
            <br />
            <ErrorBoundary showErrorMsg>
              <InstantBugginess />
            </ErrorBoundary>
          </div>
        )}

        {type === 'no-error-boundary' && (
          <div style={demoStyle}>
            <h3>
              NO BOUNDARY - <span style={{ color: 'red' }}>SUPER FAIL</span>
            </h3>
            <p className='no-line-height'>These components are NOT inside of a boundary.</p>
            <p className='no-line-height'>So if any component crashes, all other components and the page will break.</p>
            <div className='container' style={containerStyle}>
              <BuggyCounter component='header' />
              <BuggyCounter component='body' />
              <BuggyCounter component='footer' />
            </div>
          </div>
        )}
      </div>
    </div>
  )
}

stories
  .add('Default', () => getItem('default'))
  .add('With Retries', () => getItem('retries'))
  .add('Component Level', () => getItem('component-level'))
  .add('Entry Error', () => getItem('entry-error'))
  .add('With No Error Boundary', () => getItem('no-error-boundary'))

interface BuggyCounterProps {
  component: string
}

const BuggyCounter: React.FunctionComponent<BuggyCounterProps> = ({ component }) => {
  const [count, setCounter] = useState(0)

  const handleClick = () => setCounter(count + 1)
  if (count === 3) throw new Error('I crashed!')
  return (
    <div style={{ border: '1px dashed #ccc', padding: '0 15px 15px', margin: 15, borderRadius: 5 }}>
      <h3>{component}</h3>
      <button onClick={() => handleClick()}>+</button> <strong>{count}</strong>
    </div>
  )
}

const InstantBugginess = ({ message = 'This is a simulated crash!' }) => {
  throw new ReferenceError(message)
}
