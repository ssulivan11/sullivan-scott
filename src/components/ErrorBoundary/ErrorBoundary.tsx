import React from 'react'
import Alert from '../Alert/Alert'

interface Props {
  /**
   * Callback function once error retry limit is reached
   */
  callbackFunction?: () => {}
  /**
   * Set the amount of times to retry a success state until error is captured
   */
  retries: number
  /**
   * If error is caught, to show or not to show the error message
   */
  showErrorMsg?: boolean
  /**
   * If showErrorMsg, then you can define a custom error message
   */
  errorMsg?: string
  /**
   * Specify the content of your ErrorBoundary
   */
  children?: JSX.Element[] | JSX.Element
}

interface State {
  hasError: boolean
  errorCount: number
}

class ErrorBoundary extends React.Component<Props, State> {
  static defaultProps = {
    retries: 0,
  }

  constructor(props) {
    super(props)
    this.state = {
      hasError: false,
      errorCount: 0,
    }
  }

  componentDidCatch(error, errorInfo) {
    const { retries, callbackFunction } = this.props
    if (retries >= 1) {
      this.setState((oldState) => ({ errorCount: oldState.errorCount + 1 }))
    } else {
      this.setState({
        hasError: true,
      })
      if (typeof callbackFunction === 'function') callbackFunction()
      console.error(error, errorInfo)
    }
  }

  render() {
    const {
      children,
      retries,
      showErrorMsg = false,
      errorMsg = "Oops, something's gone wrong, please try again later.",
    } = this.props
    const { errorCount, hasError } = this.state

    if ((retries >= 1 && errorCount > retries) || hasError) {
      if (showErrorMsg) return <Alert kind='error' message={errorMsg} />
      return null
    }
    return children
  }
}

export default ErrorBoundary
