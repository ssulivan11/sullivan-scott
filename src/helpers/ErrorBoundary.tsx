import * as React from 'react'

interface Props {
  callbackFunction: object
  retries: number
  children: JSX.Element[] | JSX.Element
}

interface State {
  hasError: boolean
  errorCount: number
}

class ErrorBoundary extends React.Component<Props, State> {
  static defaultProps = {
    retries: 0,
    callbackFunction: () => {},
  }

  constructor(props) {
    super(props)
    this.state = {
      hasError: false,
      errorCount: 0,
    }
  }

  componentDidCatch(error, errorInfo) {
    const { retries, callbackFunction } = this.props
    if (retries >= 1) {
      this.setState((oldState) => ({ errorCount: oldState.errorCount + 1 }))
    } else {
      this.setState({
        hasError: true,
      })
      // Object.keys(errorInfo).forEach((key) => {
      //   console.warn(key, errorInfo[key])
      // })
      if (typeof callbackFunction === 'function') callbackFunction()
    }
  }

  render() {
    const { children, retries } = this.props
    const { errorCount, hasError } = this.state

    if ((retries >= 1 && errorCount > retries) || hasError) return null
    return children
  }
}

export default ErrorBoundary
