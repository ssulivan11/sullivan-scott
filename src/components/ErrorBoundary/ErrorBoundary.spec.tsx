import * as React from 'react'
import { shallow, mount } from 'enzyme'
import ErrorBoundary from './ErrorBoundary'

const defaultProps = {
  retries: 0,
  children: <div />,
  callbackFunction: jest.fn(),
}

const withCallbackFn = {
  ...defaultProps,
  retries: 1,
  callbackFunction: jest.fn(),
}

const badCallbackFn = {
  ...defaultProps,
  callbackFunction: null,
}

const errorMessage = {
  ...defaultProps,
  showErrorMsg: true,
  errorMsg: 'Oops, we got problems',
}

const MockedComponent = () => <div />

const setup = (props, renderer = shallow) => {
  spyOn(console, 'error') // suppress errors from logging
  const component = renderer(
    <ErrorBoundary {...props}>
      <MockedComponent />
    </ErrorBoundary>,
  )

  const error = new Error('mock test error')
  component.find(MockedComponent).simulateError(error)

  return {
    getComponent: () => component,
    debug: () => component.debug(),
  }
}

describe('ErrorBoundary', () => {
  it('Log Error and Build ErrorBoundary Component', () => {
    const component = setup(defaultProps, mount)
    expect(component.getComponent().props().children.type).toBeDefined()
  })

  it('Log Error with multiple retries', () => {
    const component = setup(withCallbackFn, mount)
    expect(component.getComponent().props().children.type).toBeDefined()
  })

  it('Should NOT the callback function if not a fn', () => {
    const component = setup(badCallbackFn, mount)
    expect(component.getComponent().prop('callbackFunction')).toBeNull()
  })

  it('Should NOT the callback function if not a fn', () => {
    const component = setup(errorMessage, mount)
    expect(component.getComponent()).toMatchSnapshot()
  })
})
