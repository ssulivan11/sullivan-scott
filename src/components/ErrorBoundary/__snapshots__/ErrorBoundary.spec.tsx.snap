// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Should NOT the callback function if not a fn 1`] = `ReactWrapper {}`;
